### DISCUSSION TITLE
Q/A: Would you want any z/OS 'exploitation' in tools being ported?

### DISCUSSION THREAD

**Discussion by MikeFultonDev** [zopencommunity/meta#27] [Ideas]
z/OS has a number of features unique to z/OS. How important is it that tools support these features? 

Some examples that spring to mind:
- EBCDIC support for tools like git
- certificate support in SAF from tools like git and curl

**Comment by v1gnesh** [comment]
Very; it's the exact thing that the platform should do.
Porting things without hooks into the platform seems of little use.

**Reply by jshimoda9323** [reply]
For the EBCDIC support issue, I'm personally on the fence about it.  The character encoding of files is such a fundamental attribute of a development environment that I'm afraid we would need to add EBCDIC support for all or most of the OSS tools, if we did it (i.e. if we add EBCDIC support to 'git' it would only make sense to also add it for 'less', and any diff utilties, etc).  I'm also concerned that adding such support would not only be difficult/time consuming, but also a problem for maintenance in some cases (especially for those OSS maintainers who have never dealt with EBCDIC before).  OTOH EBCDIC is a fundamental part of z/OS.  Going the ASCII only route could allow faster porting and easier maintenance in the long run.  This does beg the question - what is the strategic direction for the type of character encoding of our ported OSS tools?  Should we just support ASCII/UTF, or ASCII/UTF+EBCDIC?

**Reply by IgorTodorovskiIBM** [reply]
My preference would be to continue leveraging the enhanced ascii support. Node.js, Go, and Python for z/OS are all built in ASCII mode and it has minimized maintenance cost significantly for us.  There is the question of how we handle untagged files and we should probably have a broader discussion on that. 

One of my goals early on was to develop a "support" library that would make sure that our ports are consistent. We created zoslib for Node.js for this purpose: https://github.com/ibmruntimes/zoslib and we have certain logic as well as environment variables on how untagged files are handled.

There's also some limitations with enhanced ascii support, for example, by default pipes are untagged and therefore no translation is done between pipes unless you explicitly tag them. Developing a C runtime wrapper to wrap common functions like pipe would help further minimize the # of changes needed.

**Reply by MikeFultonDev** [reply]
Strategically it would seem UTF makes the most sense because many languages and customers demand more than a single-byte character encoding. With that in mind, ASCII/UTF is most pragmatic. However, I would say git is 'special' in that the tool itself can be ASCII/UTF but it's 'payload' (i.e. the source code in the repos) needs to be able to support EBCDIC files (of various code pages) well, since existing COBOL, PL/I, assembler is written that way and it can't necessarily be 'converted' to ASCII in a loss-less way. 

So my 'EBCDIC for git' was probably too short a statement :)

**Reply by v1gnesh** [reply]
When you say Git, do you mean Git client or Git server?
If it's client, why, considering that there's already Rocket Software's port of Git.

I would have thought it makes more sense to have Git server running on Z, with native support for zOS and EBCDIC.
With DSFS and whatever else, IBM can create a compelling case to keep Z (zOS or LoZ) _the_ source (app source code, PARMLIBs, config members, whatever). 
Not having to mirror to enterprise git hosted on distributed, not having to use workflows & tools of the distributed side (whose whole approach & reasoning may be different) is bound to be good for retaining things on Z.

Also, is ease of maintenance the only important thing, even If it's going to drive people away from the platform?
Ex: As people/orgs realise they can do things with unix-y tools on Z (ASCII/UTF8), isn't it an "obvious conclusion" for them to then think, if I can do it with just ASCII/UTF8 unix-y tools, why does it have to run on Z at all.

IMHO, proper, native integration into zOS is what will make the platform appealing, while allowing IBM to illustrate the benefits of the mainframe. Or else, why is Golang supporting VSAM files?
As people enjoy (parallel processing) the benefits of multi-core desktops etc., what better way is there to attract people than to show that they can do the same thing on Z (including the platform-exclusive bits), while benefiting from the platform's massive core count, memory & utils that natively support them.

**Reply by MikeFultonDev** [reply]
@v1gnesh The git client from Rocket software is binary only. We want to provide a true open source version of git. It would be great to have Rocket provide the source patches and build information to help on this path if they are willing. 
Is there a 'git server' ? I thought git was git? 
I agree exploitation of z/OS capabilities for Open Source software is important. I have to admit I am on step one of 'get it to work' first. 'get it to work' to me means having good support for a variety of code pages for ASCII, UTF8, EBCDIC, and maybe even DBCS.

**Reply by v1gnesh** [reply]
>  Is there a 'git server' ? I thought git was git?

I'm looking for the right way to say it then...
What I mean is, it's pointless to run git on zOS, only to also require (invisibly) cloning/syncing the same to/with a distributed git product (bitbucket, gh enterprise, etc.). 
Totally get it, first steps first; just curious if a front end for the git port will be available. 
Or if existing git products can run _in zOS_ (not in s390x, not in zCX).
Ex: I see that GitLab can run in s390x; GH Enterprise, I don't know.

**Reply by daver1958** [reply]
Compression tools should make use of the compression offload engine in zos and zlinux (and hopefully enabled in zcx one day), ditto sort, encryption etc.

**Reply by v1gnesh** [reply]
Sure Dave... have you seen how zstd compares with zlib though?
https://gregoryszorc.com/blog/2017/03/07/better-compression-with-zstandard/

**Reply by jrnewton** [reply]
> @v1gnesh The git client from Rocket software is binary only. We want to provide a true open source version of git. It would be great to have Rocket provide the source patches and build information to help on this path if they are willing. 

Rocket provides full source archives through their community portal site... All packages updated as of April 2022.  Seems like a good starting point, no?  That said, it's less than ideal that no git/patch information is provided, making it hard to tie the source to a specific point in time in the git history of each repo.

**Reply by MikeFultonDev** [reply]
Yes. Definitely it makes sense to use the updates rocket made.

**Comment by daver1958** [comment]
I have, I mostly use it because it's suited to distributed zfs that I run my docker version of zd&t on. Here is a different version of the comparisons - http://facebook.github.io/zstd/ .

**Comment by MikeFultonDev** [comment]
Another thought - for those that code in 'native' z/OS languages like COBOL, PL/I, HL/ASM...

One of the tools we are porting is 'automake'. One of the nice things it does for C code is to generate make dependencies for header files.
Is there interest in adding this capability for other languages? The source code would likely need to reside in the 'file system' (as opposed to datasets) so that there are proper time stamps.

**Comment by daver1958** [comment]
Now that we have zsh on z/os v3.1 (yay that was created in the early 90's), can we have oh my zsh and bring up uss command line to 2010's? It advertises - "Oh My Zsh will not make you a 10x developer...but you may feel like one", i need that!

**Reply by v1gnesh** [reply]
Have a look at https://github.com/ZOSOpenTools/powerlinegoport with bash for now.

I don't know about oh my zsh, but https://github.com/romkatv/powerlevel10k requires libgit2 and gitstatus porting first.

**Reply by AnthonyGiorgio** [reply]
I tried out powerlinegoport, and there were some display issues.  It's mostly there, but not quite ready for production.

**Reply by v1gnesh** [reply]
Have you tried it on your workstation, and does it work fine there?

**Reply by AnthonyGiorgio** [reply]
It works fine on my Fedora Linux machine.  In fact, I switched to it as my primary shell for now.

**Reply by v1gnesh** [reply]
:thinking: The only customization I sent was https://github.com/justjanne/powerline-go/pull/381.
